File: src/main/java/ru/ioleksiv/telegram/bot/core/controller/handler/Handler.java
package ru.ioleksiv.telegram.bot.core.controller.handler;

import ru.ioleksiv.telegram.bot.api.model.objects.Update;
import ru.ioleksiv.telegram.bot.api.model.result.HandlerResult;
import ru.ioleksiv.telegram.bot.core.controller.handler.check.Validator;
import ru.ioleksiv.telegram.bot.core.controller.handler.invoke.Invoker;
import ru.ioleksiv.telegram.bot.core.controller.handler.unpack.Unpacker;
import ru.ioleksiv.telegram.bot.core.model.ITelegram;

import java.util.Optional;

public class Handler<ARG extends ITelegram> {

    private final Invoker<ARG> methodInvoker;

    private final Validator<Update> updateValidator;

    private final Unpacker<Update, ARG> updateUnpacker;

    public Handler(Invoker<ARG> methodInvoker,
                   Validator<Update> updateValidator,
                   Unpacker<Update, ARG> updateUnpacker) {
        this.methodInvoker = methodInvoker;
        this.updateValidator = updateValidator;
        this.updateUnpacker = updateUnpacker;
    }

    public HandlerResult run(Update update) {

        return Optional.ofNullable(update)
                .map(updateUnpacker::unpack)
                .filter(Optional::isPresent)
                .map(Optional::get)
                .map(methodInvoker::run)
                .orElse(HandlerResult.pass());
    }

    public boolean hasSubscription(Update update) {
        return updateValidator.check(update);
    }

}

File: core/src/main/kotlin/io/github/oleksivio/tl/kbot/core/core/controller/handler/Handler.kt
package io.github.oleksivio.tl.kbot.core.core.controller.handler

import io.github.oleksivio.tl.kbot.server.api.objects.Update
import io.github.oleksivio.tl.kbot.core.model.result.HandlerResult
import io.github.oleksivio.tl.kbot.core.core.controller.handler.check.Validator
import io.github.oleksivio.tl.kbot.core.core.controller.handler.invoke.Invoker
import io.github.oleksivio.tl.kbot.core.core.controller.handler.unpack.UnpackerFunction
import io.github.oleksivio.tl.kbot.server.api.model.ITelegram

class Handler<ARG : ITelegram>(private val methodInvoker: Invoker<ARG>,
                               private val updateValidator: Validator<Update>,
                               private val updateUnpackerFunction: UnpackerFunction<Update, ARG>
) {

    fun run(update: Update): HandlerResult {
        val unpacked = updateUnpackerFunction(update) ?: return HandlerResult.pass()
        return methodInvoker.run(unpacked)
    }

    fun hasSubscription(update: Update): Boolean {
        return updateValidator.invoke(update)
    }

}
